docs/GDPs/gdp-0003-duck-storage.rst

```diff
@@ -0,0 +1,822 @@
+    are removed from :code:`axes` in the returned Storage, while slices do not reduce
+    dimensionality.
+
+    Parameters:
+        + :code:`key: index_like` Indicates the indices from which the data of the storage is to be
+          returned. The same keys as in
+          `NumPy Indexing <https://numpy.org/doc/stable/reference/arrays.indexing.html>`_
+          are allowed, with the addition that
+
+          + keys can be :code:`cp.ndarrays` whenever a :code:`np.ndarray` is valid or
+
+          + a boolean or integer GT4Py storage with the same shape as :code:`self` (possibly with
+            permuted axes).
+
+
+:code:`__setitem__(self: Storage, key: key_like, Value) -> None`
+    Set the data of the storage at a certain index, in a subregion or
+    at selected locations of the underlying buffer
+
+    The :ref:`broadcasting` behavior and device selection are equivalent to that of a unary ufunc
+    with a provided output buffer. For example, :code:`stor_out[key] = stor2d` would be
+    equivalent to :code:`np.positive(stor2d, out=stor_out[key]`)
+
+    Parameters:
+        + :code:`key: index_like` Indicates the locations at which the values are to be changed. The
+          same keys as for :code:`__setitem__` are supported.
+
+        + :code:`value: Union[Number, Storage, cp.ndarray, np.ndarray]` the values that are copied
+          to the storage at the locations indicated by :code:`key`.
+
+:code:`copy(self: Storage) -> Storage`
+    Create a new Storage instance with the same parameters as this instance and a copy of the data.
+
+:code:`reinterpret(self, axes) -> Storage`
+    Parameters:
+        + :code:`axes: Sequence[str]` the axes of the resulting storage
+
+    A view of the buffer, with the axes relabeled according to :code:`axes`. The behavior seems
+    similar to transpose, however the shape and strides remain unchanged, meaning that not only the
+    order of indexing changes, but also the semantic meaning attached to the data. For example, if
+    :code:`self` is an IJK-storage with :code:`shape==(10,20,30)`, and we call
+    :code:`self.reinterpret("KJI")`, a KJI-storage with the same shape as before is returned,
+    meaning that e.g. the size in the K-dimension will change from 30 to 10 in the result.
+
+:code:`to_cupy(self: Storage) -> cp.ndarray`
+    Return a view of the underlying device buffer (CuPy :code:`ndarray`) if present or raise an
+    exception if this instance does not contain a device buffer.
+
+:code:`to_ndarray(self: Storage) -> Union[np.ndarray, cp.ndarray]`
+    Return a view of the device buffer (CuPy :code:`ndarray`) if present or a view of the host
+    buffer (NumPy :code:`ndarray`) otherwise.
+
+:code:`to_numpy(self: Storage) -> np.ndarray`
+    Return a view of the underlying host buffer (NumPy :code:`ndarray`) if present or raise an
+    exception if this instance does not contain a host buffer.
```

Comment:
Thanks! Could an exception class be made for the not-present case and documented? I could see wanting to catch such an exception.

Reply:
proposed `GTNoSuchBufferError`

Reply:
Is it necessary to prepend it with `GT`, since it's already a `gt4py.NoSuchBufferError`?

Reply:
Other exceptions are prefixed `GTScript`, at least there would be some consistency.